ENH: Push histogram calculations to compiled_base

theodoregoetz opened the pull request:
# Faster numpy.histogram() and histogramdd()

The **fast-histogram** python package demonstrates that histogramming real data onto a continuous range with evenly-spaced bins, can be made significantly faster. I took this idea of pushing the bin calculation and the filling of the histogram array into a C function and implemented it in **NumPy**.

For details, please my gist concerning this patch:
https://gist.github.com/theodoregoetz/10d2351421689bf2660b4f2fca350e6e

ahaldane committed:
MAINT: improve struct assignment, allow multi fields views

This commit attempts to make structure assignment more consistent, and
then changes multi-field indices to return a view instead of a copy.

Assignment between structures now works "by field position" rather than
"by field name".

Fixes #2353, fixes #6085, fixes #3351, fixes #6085, fixes #6314,
fixes #2346, fixes #7058, fixes #3641, fixes #5994, fixes #7262,
fixes #7493

ahaldane committed:
TST: Fix/Remove old unit tests for structure assignment

ahaldane committed:
TST: Add new unit tests for structure assignment

emilienkofman committed:
BUG: Fixes #7395, operator.index now fails on numpy.bool_

perimosocordiae committed:
ENH: add hermitian=False kwarg to matrix_power

With a symmetric matrix, the more efficient `eigvalsh` method
can be used to find singular values.

wirawan0 committed:
DOC: Clarifying the meaning of small values for `suppress` print option.

eric-wieser committed:
BUG: Fix pinv on matrix stacks

Fixes #8826

eric-wieser committed:
MAINT: Remove use of int as boolean

eric-wieser committed:
ENH: Broadcast the second argument of matrix_rank to match pinv

eric-wieser committed:
BUG: Prevent hang when sorting dtype=S0 arrays

eric-wieser committed:
BUG: Prevent ValueError when resizing V0 arrays

eric-wieser committed:
BUG: Allow empty strings to be viewed as themselves

eric-wieser committed:
BUG: Allow empty strings to be pickled

eric-wieser committed:
MAINT: Pre-emptively handle the case when elsize == 0

Fixes np.fromiter and np.dtype. Neither of these code paths can currently be hit.

eric-wieser committed:
BUG: Allow array with 0 dims to be resized

ahaldane committed:
TST: update tests to remove unneeded whitespace

eric-wieser committed:
DEP: 0 should be passed to bincount, not None

Fixes gh-8841

eric-wieser committed:
MAINT: Make it clear how to remove the deprecation

ahaldane committed:
ENH: remove unneeded spaces in float/bool reprs

eric-wieser committed:
ENH: Add out argument to `concatenate`

eric-wieser committed:
MAINT: Extract duplicated priority code

b-carter committed:
TST: add test for unsigned bins monotonicity check, see #9222

b-carter committed:
BUG: fixes unsigned bins monotonicity check, see #9222

b-carter committed:
minor change to the logic

ahaldane committed:
BUG: PyArray_CountNonzero does not check for exceptions

Fixes #9295

WarrenWeckesser committed:
BUG: f2py: Convert some error messages printed to stderr to exceptions.

In f2py/src/fortranobject.c, the function `check_and_fix_dimensions` does
pretty much what it says on the tin:
    /*
      This function fills in blanks (that are -1\'s) in dims list using
      the dimensions from arr. It also checks that non-blank dims will
      match with the corresponding values in arr dimensions.
    */
There are several error conditions detected by the function.  In the code
before this change, when the function detected such an error, it would
print a message to stderr and return 1.

In this change, when an error is detected in `check_and_fix_dimensions`,
an exception is set.

This new feature of `check_and_fix_dimensions` is used in three places
in the function `array_from_pyobj()`.
In each case, there was an old comment of the form:
    /* XXX: set exception */
In this change, the exception is now set in `check_and_fix_dimensions()`,
so those comments have been removed.

For some error conditions, the new code changes the exception that is
raised.  For example, here's a scipy test before this change:

```
In [5]: from scipy.linalg import _fblas as fblas

In [6]: a = np.array([[1., 0.], [0., -2.], [2., 3.]])

In [7]: b = np.array([[0., 1.], [1., 0.], [0, 1.]])

In [8]: fblas.dsyr2k(a=a, b=b, alpha=1.0, c=np.zeros((15, 8)))
0-th dimension must be fixed to 3 but got 15
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-8-bc4d51f8d016> in <module>()
----> 1 f(a=a, b=b, alpha=1.0, c=np.zeros((15, 8)))

error: failed in converting 2nd keyword `c' of _fblas.dsyr2k to C/Fortran array
```

After this change, we get:

```
In [2]: from scipy.linalg import _fblas as fblas

In [3]: a = np.array([[1., 0.], [0., -2.], [2., 3.]])

In [4]: b = np.array([[0., 1.], [1., 0.], [0, 1.]])

In [5]: fblas.dsyr2k(a=a, b=b, alpha=1.0, c=np.zeros((15, 8)))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-bc4d51f8d016> in <module>()
----> 1 f(a=a, b=b, alpha=1.0, c=np.zeros((15, 8)))

ValueError: 0-th dimension must be fixed to 3 but got 15
```

The spurious print has been changed to the exception message, but some
potentially useful information in the old exception message has been
lost.

ElieGouzien committed:
DOC: Add examples for np.arg[min|max|sort]

Show how to retrieve indices for d-dim arrays.

[ci skip]

ahaldane committed:
ENH: "pad_sign" option, also threshold and edgeitems

eric-wieser committed:
ENH: Make duplicated masked constants obvious

eric-wieser committed:
BUG: np.ma.masked does not preserve identity through pickle

It's still possible to create duplicate MaskedConstants through `.copy()`

eric-wieser committed:
BUG: Prevent copies of np.ma.MaskedConstant from being created by .view.

This seems to solve the problem everywhere. It's not clear if this works on PyPy.

Fixes gh-9328

eric-wieser committed:
BUG: np.ma.masked is mutable

eric-wieser committed:
MAINT: improve wording of error messages

eric-wieser committed:
ENH: Add the out parameter to stack too

ahaldane committed:
ENH: Simplify some code in arrayprint.py

eric-wieser committed:
BUG: Allow 0d arrays instead of scalars in gradient

This fixes gh-8292

eric-wieser committed:
BUG: Only allow 1d distance arrays

2d arrays would work, but in unpredictable and undocumented ways.

This at least makes gh-9401 give a better error message.

eric-wieser committed:
BUG: Use np.ndim not asarray, to allow duck-types

eric-wieser committed:
MAINT: Use clearer variable

eric-wieser committed:
MAINT: Use zip, not enumerate

Double-indexing is just a little harder to read

eric-wieser committed:
MAINT: Improve comments and tests, in response to review

eric-wieser committed:
MAINT: Collect together _arraymethods, and remove duplicate 'copy'

eric-wieser committed:
BUG: Keep mask when calling methods on 0d arrays.

This also prevents `.squeeze()` from ever returning `masked`, which was incorrect

Fixes gh-9430.

eric-wieser committed:
BUG: Make transpose and diagonal masks be views

Fixes gh-8506 and fixes gh-7483

eric-wieser committed:
BUG: change fix for gh-9328 to make masked.copy() a noop

eric-wieser committed:
TST: Add test for direct writing of `masked`

xoviat committed:
distutils: gnu: patch fcompile

This allows mingw's gfortran to work with MSVC. DLLs are
autogenerated using heuristics that should work with most
cases. In addition, a libopenblas DLL is compiled from
the static lib for use with MSVC.

All generated DLLs have randomized names so that no clashes
will occur.

eric-wieser committed:
DEP: Letting fromstring pretend to be frombuffer is a bad idea

Interpreting a unicode string as raw binary data is a terrible idea, especially if the encoding is determined by the system (python 2)

pv committed:
distutils: handle unlinkable object files in build_clib/build_ext, not gnu

Add concept of unlinkable Fortran object files on the level of
build_clib/build_ext.

Make build_clib generate fake static libs when unlinkable object files are
present, postponing the actual linkage to build_ext.

This enables MSVC+gfortran DLL chaining to only involve those DLLs that
are actually necessary for each .pyd file, rather than linking everything
in to every file. Linking everything to everywhere has issues due to
potential symbol clashes and the fact that library build order is
unspecified.

Record shared_libs on disk instead of in system_info. This is necessary
for partial builds -- it is not guaranteed the compiler is actually called
for all of the DLL files.

Remove magic from openblas msvc detection. That this worked previously
relied on the side effect that the generated openblas DLL would be added
to shared_libs, and then being linked to all generated outputs.

xoviat committed:
distutils: system_info: fixup

xoviat committed:
distutils: gnu: hash files + chained dlls

xoviat committed:
distutils: gnu: revert :art:

xoviat committed:
distutils: gnu: cleanup

xoviat committed:
distutils: gnu: revert :art:

ahaldane committed:
MAINT: array_any_nonzero now uses new iterator API

eric-wieser committed:
DEP: Deprecate the event argument to datetime types, which is ignored since 1.7.0

solarjoe committed:
DOC: Improve ndarray.shape property documentation.

[ci skip]

b-carter committed:
allow non-strictly increasing bins

bashtage committed:
ENH: Remove unnecessary restriction in noncen-f

Set dfnum restriction to be > 0 as required by noncentral chisquare

closes #6638

srinivasreddy committed:
MAINT: Remove unnecessary imports

- Now that `set` datastructure has been added the `builtin` module, we do not need to explicitly import from `sets` module

irushchyshyn committed:
BUG: fix infinite loop when creating np.pad on an empty array

MSeifert04 committed:
DOC: rot90 wrongly positioned versionadded directive. [skip ci]

srinivasreddy committed:
MAINT: Reuse the code to compute sha256, md5 hashes

srinivasreddy committed:
MAINT: Remove global statement

- Use of `global` variable may have unintented consequences, so it is better to refactor the code
so that we do not need to use `global`.

srinivasreddy committed:
BUG: add `__ne__`  method to dummy_ctype class.

There is no Python defined relation between `__eq__` and `__ne__`,  so it is
recommended that both be defined.

pv committed:
BUG: core: fix wrong method flags for scalartypes.c.src:gentype_copy

pv committed:
BUG: core: fix array_copy_keeporder METH_VARARGS signature

charris committed:
MAINT: Different fix for scalartypes.c.src `__copy__` method.

Create a new function `scalartype.c.src:gentype___copy__` that does not
take keyword arguments. This is done by adding `__copy__` to the same
code generator loop that generates `gentype___deepcopy__`.

srinivasreddy committed:
MAINT: remove try..except clause.
- Generator type has been added to types module in 2.2 we do not need to catch AttributeError

theodoregoetz committed:
ENH: push histogram calculations to compiled_base

The fast-histogram python package demonstrated that
histogramming real data onto a continuous range with
evenly-spaced bins can be made significantly faster.
Two methods were added to multiarray/compiled_base.c:
_arr_histogram_uniform(x, edges, weights) and
_arr_histogramdd_uniform(x, edges, weights). These
methods make the assumption that the edges are
uniform-linear, i.e. the result of np.linspace() or
similar. The ensurance that this is case is handled on
the python side in np.histogram() and np.histogramdd().
Benchmarks were added for 1D and 2D histogramming for
when the histogram spans the range of the sample and
when it spans only 1% of the range of the sample (per
dimension).

charris committed:
DOC: Update release notes for noncentral_f changes.

The `dfnum` parameter is now required to be > 0 rather than
the previous > 1.

[ci skip]

MSeifert04 committed:
MAINT: Fix a comment regarding the formula for arange length [skip ci]

bashtage committed:
DOC: Fix type definitions in mtrand

Improve description of dtype requirements to correctly indicate float is
supported
Add required range

[skip ci]

nschloe committed:
f2py: allow Fortran arrays of dimension 0

Up until now, f2py throw an error when arrays were declared which had
dimension of length 0. This, however, is a perfectly legal case, and in
fact occurs frequently in the context of linear algrebra. This bug was
discovered, for example, in an interface that does matrix
tridiagonalization. If the matrix is 1x1, the super- and subdiagonal are
of length 0.

Note that negative dimensions continue to produce errors although
Fortran also allows this (it always allocates to size max(0, n)).

nschloe committed:
f2py: rename internal function

nschloe committed:
[DOC] add release note for f2py's dimension arrays

nschloe committed:
[TST] add tests for 0-sized Fortran arrays

nschloe committed:
[TST] fix test for 0-sized array operations

nschloe committed:
[TST] slightly simplify f2py test

ExpHP committed:
BUG: memory leak in np.dot of size 0

ExpHP committed:
TST: test for leak in np.dot of size 0 arrays

charris committed:
BUG: Fix broken runtests '-t' option.

The '-t' was broken when init files were added to the tests directories,
as it was no longer valid to import tests from the numpy git repo. The
fix here is the use a path to the installed numpy. Which numpy that is
depends on the '--no-build' option.

[ci skip]

theodoregoetz commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -278,7 +278,7 @@ arr_histogramdd_uniform(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kw
         PyErr_SetString(PyExc_TypeError, "Couldn't parse the input edges.");
         goto fail;
     }
-    npy_intp nedges_total = 0;
```
This was fine. Appveyer's error message was misleading. Travis-CI caught the actual bug which was fixed in commit cb117ec194a9073e4cbfa78800fb7d6edd8bb339

jrbourbeau committed:
Updates order of parameters in save docstring to match function parameter order

mattip committed:
BUG: test, fix issue #9620 __radd__ in char scalars

juliantaylor committed:
MAINT: restore compiler warning check to zero warnings

juliantaylor committed:
MAINT: fix implicit declaration warning

juliantaylor committed:
MAINT: bump 32 bit test to zesty

juliantaylor committed:
MAINT: add compiler warning tests to all jobs

Except the builtin lapack job as that is mostly generated code.

juliantaylor committed:
MAINT: fix std header include order

Removes _POSIX_C_SOURCE redefine warnings. Standard library headers
should always be included last.

juliantaylor committed:
MAINT: remove -Wsign-compare from warnings

Too noise and not really worth fixing.

juliantaylor committed:
MAINT: merge python -3 and -OO test

They are not worth two seperate slow test runs.

juliantaylor committed:
BUG: ensure consistent result dtype of count_nonzero

The slowpath using apply_along_axis for size 1 axis did not ensure that
the dtype is intp like all other paths. This caused inconsistent dtypes
on windows where the default integer type is int32.
Closes gh-9468

irushchyshyn committed:
BUG: fix padding an empty array in reflect mode.

Check that axes with non-zero padding are non-empty.

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+    max = e[bins];
+
+    /* Possible divide-by-zero here */
+    norm = bins / (max - min);
```
`norm <= 0` should throw an error

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+    arr_x = (PyArrayObject *)PyArray_ContiguousFromAny(obj_x, NPY_DOUBLE, 1, 1);
+    if (arr_x == NULL) {
+        PyErr_SetString(PyExc_TypeError, "Couldn't parse the input array.");
```
when array constructors fail they already set an error message, I don't think extra information is needed here.

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+    arr_edges = (PyArrayObject *)PyArray_ContiguousFromAny(obj_edges, NPY_DOUBLE, 1, 1);
+    if (arr_edges == NULL) {
+        PyErr_SetString(PyExc_TypeError, "Couldn't parse the input edges.");
```
same double error setting

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+    if (weights == Py_None) {
+        hist = (PyArrayObject *)PyArray_ZEROS(1, &bins, NPY_INTP, 0);
+        if (hist == NULL) {
+            PyErr_SetString(PyExc_ValueError, "Couldn't build output array");
```
same double error setting

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+    else {
+        arr_w = (PyArrayObject *)PyArray_ContiguousFromAny(weights, NPY_DOUBLE, 1, 1);
+        if (PyArray_SIZE(arr_w) != len) {
+            if (!PyErr_Occurred()) {
```
when an error occurred PyArray_Size already crashed with null dereference, needs another array creation check again

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+        hist = (PyArrayObject *)PyArray_ZEROS(1, &bins, NPY_DOUBLE, 0);
+        if (hist == NULL) {
+            PyErr_SetString(PyExc_ValueError, "Couldn't build output array");
```
same double error setting

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+        goto fail;
+    }
+    if (PyArray_DIM(arr_bins, 0) != ndim) {
+        if (!PyErr_Occurred()) {
```
this check seems unnecessary

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+        e += bins[d] + 1;
+    }
+    e = NULL; /* prevent accessing invalid address */
```
putting e into a smaller scope for this loop is better

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,349 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+        max[d] = e[bins[d]];
+
+        /* Possible divide-by-zero here */
+        norm[d] = bins[d] / (max[d] - min[d]);
```
again imo it makes sense to just raise an error here

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -821,17 +1164,17 @@ arr_interp_complex(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
                     const npy_double inv_dx = 1.0 / (dx[j+1] - dx[j]);
                     dres[i].real = (dy[j+1].real - dy[j].real)*(x_val - dx[j])*
-			inv_dx + dy[j].real;
+            inv_dx + dy[j].real;
```
this indentation does not look correct, probably wrong tab to space conversion

juliantaylor reviewed (changes requested):
looks generally quite good % some repeating style issues
some benchmarks for our suite would be nice too, see the `benchmarks/` folder

juliantaylor commented on numpy/lib/function_base.py:
```diff
@@ -743,40 +742,18 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
                 tmp_w = weights[i:i + BLOCK]
 
-            # Only include values in the right range
             keep = (tmp_a >= mn)
```
don't know how relevant this is, but can't this also be moved into the C code?

theodoregoetz replied:
I left this out originally, but found that when the histogram covers only a small part of the sample range (10% or less), adding this back made the calculation much faster (well more than a 2x speed up) and it did not change the timing much for when the histogram range was wider than the sample range. (I'm in the process of adding benchmarks for these cases.)

juliantaylor commented:
the commit prefix for this should be ENH: not MAINT

ahaldane committed:
MAINT: Simplify array_contains using PyArray_Any

theodoregoetz commented:
@juliantaylor I believe I've addressed all comments of your first review. Should I squash the commits on my fork? (I will change the commit header to ENH then)

theodoregoetz commented:
I added benchmarks for np.histogram and np.histogramdd as well. They are not much but they cover the cases I found to be important for benchmark considerations. I am not sure if just adding the classes and test methods was enough or if I should make a note/change elsewhere in the repo.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -743,40 +742,18 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
             else:
                 tmp_w = weights[i:i + BLOCK]
 
-            # Only include values in the right range
```
Why was this comment removed?

theodoregoetz replied:
oversight. will fix.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -949,63 +928,70 @@ def histogramdd(sample, bins=10, range=None, normed=False, weights=None):
-    # Handle empty input.
-    if N == 0:
-        return np.zeros(nbin-2), edges
+        # Handle empty input.
```
Might be tidier to package this into `_histogramdd_nonuniform`, if possible

theodoregoetz replied:
I would rather make that a separate effort/pull request. I intend to investigate/profile nonuniform binning to see if a speed up can be gained over using bincount as is done here.

eric-wieser replied:
I'm not saying you should write that function in c - just break the current python code out into a sub function, so it reads vaguely like:


```python
if uniform:
    return _histogramdd_uniform(...)
else:
    return _histogramdd_nonuniform(...)
```

theodoregoetz replied:
understood. Though I'd still like to defer this to a following effort.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+        goto fail;
+    }
+    if (PyArray_DIM(arr_bins, 0) != ndim) {
+        if (!PyErr_Occurred()) {
```
What's this guarding against?

theodoregoetz replied:
I think this check is not doing anything. But I am still learning the numpy API. Will remove it if you don't object further.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+        nedges_total += bins[i] + 1;
+    }
+    if (PyArray_SIZE(arr_edges) != nedges_total) {
+        if (!PyErr_Occurred()) {
```
This one also is questionable

theodoregoetz replied:
will remove the check

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+    {
+        double *e = (double *)PyArray_DATA(arr_edges);
+        for (d = 0; d < ndim; d++) {
```
Why not move the declaration of `d` inside this block?

theodoregoetz replied:
I was following the style established in bincount just below these histogram functions where all variables are declared up top. I would be more than happy to move declarations to the tightest scope if numpy's coding style has changed since arr_bincount() was written.

juliantaylor replied:
Everything declared at the top is a bad style still present in old code, you do not have to stick to that. Declaring in the smallest scope is usually much cleaner.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+            overflow = 0;
+            {
+                double *e = (double *)PyArray_DATA(arr_edges);
+                for (d = 0; d < ndim; d++) {
```
Same for this `d`

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+            overflow = 0;
+            {
+                double *e = (double *)PyArray_DATA(arr_edges);
+                for (d = 0; d < ndim; d++) {
```
And this one. Same for `ii` and `tx` too, which could be declared within the loop

rgommers committed:
DOC: add new steering council members.

Discussion on the numpy-discussion mailing list on July 21-25 2017.

[ci skip]

pv committed:
DOC: release: add mention of msvc+gfortran to release notes

xoviat committed:
TST: appveyor: Enable OpenBLAS via MinGW Gfortran

theodoregoetz commented:
@juliantaylor @eric-wieser I think I've addressed and committed changes from your comments so far. Awaiting further comments.

theodoregoetz commented:
wow, didn't realize I had to C90 that code. Working on that now...

ahaldane committed:
ENH: release GIL in CountNonzero trivial loop

ahaldane committed:
BUG: account for field titles in multi-field indexes

Fixes #9625

MSeifert04 committed:
DOC: Correct the signature in pad doc for callable mode.

[skip ci]

MSeifert04 committed:
DOC: Fixed doctest of apply along axis with 3D return

[skip ci]

theodoregoetz commented:
@juliantaylor @eric-wieser reorganization of variable declarations done. Ready for further review.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,357 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+
+    if (bins <= 0 || (max - min) <= 0)
+    {
+        PyErr_SetString(PyExc_ValueError, "Bin edges must be increasing");
```
This error message isn't right for `bins <= 0`

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -80,6 +80,357 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+    return (PyObject *)hist;
+
+fail:
+    Py_XDECREF(hist);
```
This line will never decref - but maybe it's better to leave it here in case that changes in future

theodoregoetz replied:
good catch. and I agree leaving it in does no harm and might do good later.

xoviat committed:
TST: fix failures:

1. fail tests related to DLL load failure as they were previously untested.
2. fix have_compiler to return false on old compilers
3. xfail some tests that were not working on old Python versions.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -299,7 +304,14 @@ arr_histogramdd_uniform(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kw
 
-            if (bins[d] <= 0 || (max[d] - min[d]) <= 0)
+
+            if (bins <= 0)
```
Lost a `[d]`

theodoregoetz replied:
Turns out the public interface will never get there so I didn't add a unittest for this. The check is potentially valuable in the future so I'd leave it in. Is that OK?

lamby committed:
Make system_info output reproducible (#9652)

Whilst working on the Reproducible Builds effort [0], we noticed
that python-numpy generates output that is not reproducible. This
affects "downstream" packages such as numexpr.

This occurs as it encodes the absolute pathname into the generated file.

 [0] https://reproducible-builds.org/

Signed-off-by: Chris Lamb <chris@chris-lamb.co.uk>

charris committed:
API: Make ``a.flat.__array__`` return a copy when ``a`` non-contiguous. (#9447)

* API: Make ``a.flat.__array__`` return copy if ``a`` non-contiguous.

Previously an UPDATEIFCOPY array was returned when `a` was
non-contiguous and writeable. Exposing that type of array does not work
well with PyPy as there is no refcount, and hence the writeback never
occurs. The copy in this case is set to non-writable to expose cases
where a writeback was expected. At some future time the copy will be
made writable. See gh-7054 for a discussion of this issue. It is not
expected that this change will affect many people.

MSeifert04 committed:
BUG: Fix usage of keyword "from" as argument name for "can_cast".

Also removed inconsistency between the second argument name
between documentation ("totype") and code ("to").

theodoregoetz commented:
@juliantaylor @eric-wieser Thank you so much for your notes/review. If you have no further comments or suggestions, I can rebase and squash this pull request (changing the commit header to ENH as requested).

ahaldane committed:
BUG: remove voidtype-repr recursion in scalartypes.c/arrayprint.py

Fixes #9345

njsmith committed:
DOC: clarify wording in tutorial

In gh-9664, a user got confused by this wording, because they thought
it was talking about the length of a vector as a distance in cartesian
space versus the length of a vector as the number of elements it
contains.

njsmith committed:
DOC: Simplify terminology

For some reason this file insisted on using multiple redundant terms
for array dimensionality, and in particular liked the word "rank",
which is confusing and very rarely used in my experience. This commit
drops the word rank and removes a number of parenthetical "X (which is
to say Y)" phrasings.

xoviat committed:
appveyor: update documentation

xoviat committed:
tests: core: update failure message

ahaldane committed:
BUG: dot/matmul 'out' arg should accept any ndarray subclass

Fixes #9641

xoviat committed:
TST: utils: fix test_error_msg

Allow the error message to contain "large" arrays.

xoviat committed:
TST: util: update documentation

xoviat committed:
TST: common: update documentation

xoviat committed:
TST: block docstring: update comment

xoviat committed:
TST: callback: update comment

xoviat committed:
TST: util: fix test_error_message

use the correct replace() order

xoviat committed:
:art:

xoviat committed:
TST: utils: fix string continuation

xoviat committed:
appveyor: update documentation

yangli2 committed:
BUG: Fix possibly undefined cast of double -> long.

Current double to long casting in the zipf function depends on
non-standardized behavior when the double is too big to fit in a long.
This is potentially dangerous and makes the code fail with tools such as
AddressSanitizer.

Checks are added here to prevent overflow during casting and make sure
we get the desired behavior.

j-towns committed:
[MAINT] Add block benchmarks

Note I've parameterised the benchmarks to run at three sizes.

MSeifert04 committed:
DOC: Add whitespace after "versionadded::" directive
so it actually displays in the docs

[skip ci]

MSeifert04 committed:
DOC: Fix inline "math" directives [skip ci]

charris committed:
STY: Add blank line for PEP8.

[ci skip]

MSeifert04 committed:
DOC: Add polyutils subpackage to reference documentation [skip ci]

eric-wieser committed:
DOC: Add release notes warning about np.ma.masked immutability

jzwinck committed:
ENH: Align data in np.save() at 64 bytes (#9025)

Previously, saving format version 1 would align to 16 bytes,
and saving version 2 would align improperly (bug #8085).

Alignment is now always at least 64 bytes in either version,
which supports memory mapping of the saved files on Linux,
where mmap() offset must be a multiple of the page size.

Why 64 bytes?  Simply because we don't know of a case where
more is needed.  AVX alignment is 32 bytes; AVX-512 is 64.

Fixes #8085, closes #8598.

eric-wieser committed:
TST: Add test for operator.index(np.bool_)

ahaldane committed:
MAINT: rework recursive guard to keep array2string signature

ahaldane committed:
MAINT: arrayprint.py formatters no longer require ravel'd data

abrodkin committed:
ENH: Add support of ARC architecture

Synopsys DesignWare ARC Processors are a family of 32-bit CPUs
which in some configurations may run Linux kernel and full stack
of applications on top of it. Among other things is Python.

With that change we allow building and usage of Numpy on ARC.

xuhdev committed:
arange: `start` is not needed even when `step` is given.

The documentation says `start` must be given if `step` is specified, but this is not true. For
example, arange(5, step=2) works.

MSeifert04 committed:
DOC: Add mandatory memo argument to __deepcopy__ method documentation

[skip ci]

MSeifert04 committed:
DOC: Add keepdims argument for ndarray.max documentation

[skip ci]

eric-wieser committed:
DEP: Warn when using bool.__index__, rather than removing it completely

eric-wieser committed:
DOC: Warn about the difference between np.remainder and math.remainder

Python 3.7 adds its own `math.remainder` with different semantics to ours.
We can't change the semantics now, but we can warn users that they are different.

rmcgibbo committed:
DOX: Fix mistaken word in nanprod docstring (#9703)

eric-wieser committed:
ENH: Warn when loading old datetime pickle files

If these files were using the event feature, then there has been a silent behavioural change.
This warns the user about it.

oleksandr-pavlyk committed:
MAINT: When delta is a NumPy scalar, do multiplication in-place

This is faster and more memory efficient.

Out-of-place operation was preventing me from using linspace
to create over 2**31 equispaced doubles on my machine due to
insufficient memory.

Commit:
DOC allclose doesn't require matching shapes

Closes https://github.com/numpy/numpy/issues/5654

theodoregoetz commented:
@juliantaylor @eric-wieser @charris It's been a couple weeks so I'm checking in to see if you all think this PR looks good and is ready to merge. It is still showing "requested changes" from @juliantaylor but I believe these have all been addressed.

Commit:
Add equals to See Also; remove 'np.'

charris committed:
BUG: Make scalar function elision check writeable.

Add checks for writeable and updateifcopy in the can_elide_temp_unary
function.

Closes #9679.

dhermes committed:
Fixing typo "Porland" -> "Portland" in `building` doc.

ahaldane committed:
MAINT: fixups to legacy spacing in FloatFormat

Commit:
[ci skip] requested updates

[ci skip]

Commit:
DEP: Deprecate truth testing on empty arrays

Fixes #9583

eric-wieser committed:
MAINT: Remove unnecessary special-casing of scalars in isclose

This means that this returns an `np.bool_` instead of a `bool`, but that seems more sensible anyway.

The code was likely written this way for when scalar boolean indices didn't work

jjhelmus committed:
BUG: adjust gfortran version search regex

Adjust the gfortran regex to require a '.' to be present in the version number.
This avoids matching a hash or text containing a '-' such as the version string
from the compilers produced by crosstool-NG.  For example:
GNU Fortran (crosstool-NG 8a21ab48) 7.2.0

ahaldane committed:
MAINT: cleanup circular import b/w arrayprint.py,numeric.py

Moves array_repr, array_str, g/set_printoptions to arrayprint.py

NelleV committed:
ENH Better error message for savetxt when X.ndim > 2

savetxt does not support saving arrays of dimension 0 or higher than 2.
This pull request improves the message of the error raised.

charris committed:
BUG: Ensure `_npy_scaled_cexp{,f,l}` is defined when needed.

The `_npy_scaled_cexp{,f,l}` functions were previously only defined when
the npy_cexp fallback functions were needed, but they are also called by
the npy_csinh and npy_ccosh functions.

charris committed:
MAINT: Use PyArray_CHKFLAGS in more places.

This is more direct than `PyArray_FLAGS(a) & AFLAG`.

pv committed:
BUG: core: use npy_cabs for abs() for np.complex* scalar types

Use npy_cabs to compute absolute values for complex scalar types,
instead of reimplementing it in a buggy way.

eric-wieser committed:
BUG: Fix loss of precision for large values in long double divmod

eric-wieser committed:
BUG: Errors thrown by 0d arrays in setitem are silenced and replaced

Fixes gh-8461

MSeifert04 committed:
DOC: Fix ndarray.__setstate__ documentation, it only takes one argument.

[skip ci]

charris committed:
MAINT: Use single backticks when link needed.

[ci skip]

charris committed:
BUG: Cython 0.27 breaks NumPy on Python 3.

The latest Cython came out and it causes the cythonized mtrand.c source
to issue a compile warning for Python 3, breaking NumPy travis tests. As
a temporary fix, force the tests to use Cython 0.26.

eric-wieser committed:
TST: Check if precision is lost in longcomplex

Related to gh-8903.

eric-wieser committed:
BUG: Fix silent loss of clongdouble precision

This was not caught in gh-8903

charris committed:
MAINT: Clarify of the suppress print option more.

eric-wieser committed:
MAINT: Use the PyArray_(GET|SET)_ITEM functions where possible

eric-wieser committed:
MAINT: Remove hack from gh-7659 for gh-7493

Working with 0d arrays is enough here

eric-wieser committed:
MAINT: Remove special casing of 0d in MaskedArray.__str__

eric-wieser committed:
MAINT: Remove special casing of structured dtypes in MaskedArray.__str__

This means that large void arrays are now truncated as they are for other types, for speed.

charris committed:
MAINT,BUG: Fix mtrand for Cython 0.27.

The `import_array()` macro, that defined a C code block that included a
return, was not handled correctly. The fix here is to cdef a replacement
`import_array` function with a defined error return. The new function is
a slight variation of the corresponding function defined by Cython.

dgasmith committed:
BUG: Fixes optimal einsum path for multi-contraction intermediates

eric-wieser committed:
BUG: can_cast(127, np.int8) is False

Fixes #5917

eric-wieser committed:
BUG: np.ma.trace gives the wrong result on ND arrays

Fixes #5560

bmwiedemann committed:
ENH: Allow to override build date

and use gmtime in order to be independent of timezone
in order to make builds reproducible.
See https://reproducible-builds.org/ for why this is good
and https://reproducible-builds.org/specs/source-date-epoch/
for the definition of this variable.

bmwiedemann committed:
MAINT: remove outdated entries in f2py

because svn is not used anymore

charris committed:
DOC: Update after NumPy 1.13.2 release.

* Add 1.13.2-changelog.rst
* Add doc/release/1.13.2-notes.rst
* Update doc/source/release.rst

[ci skip]

eric-wieser committed:
BUG: Shrinking the mask on a structured type errors

Now it is a no-op, as elsewhere in the file. Extract the mask-shrinking into a function.

eric-wieser committed:
BUG: Don't ignore mismatching shapes just because the mask is all zeros

Fixes #4520

pv committed:
TST: linalg: add basic smoketest for cholesky

jarrodmillman committed:
DOC: Clean up README

lesteve committed:
DOC: add missing underscore in set_printoptions

pv committed:
CI: set correct test mode for appveyor

jarrodmillman committed:
MAINT: Add appveyor badge to README

jarrodmillman committed:
MAINT: Make appveyor config a dot-file

AppVeyor supports dot-file-style YAML named ``.appveyor.yml``:
- https://www.appveyor.com/docs/build-configuration/#yaml-file-alternative-naming

The Travis CI config is already a dot-file.

eric-wieser committed:
MAINT: Remove dead code

This could never be negative anyway, since elsize >= 0, and we already check for items >= 0.

eric-wieser committed:
BUG: Free dtype and dim objects correctly.

Previously `newdescr` was often leaked

eric-wieser committed:
DOC: Add release notes for 0-size dtypes

nafest committed:
BUG: fix stray comma in _array2string

_array2string unconditionally added a comma after the '...' that
are inserted for arrays exceeding the size threshold. Instead, add
the separator character in _formatArray. Fixes #9777.

charris committed:
DOC: Update ndarray.shape property documention.

[ci skip]

xoviat committed:
DOC: update scipy.integrate recommendation

Closes scipy/scipy#7965.

xoviat committed:
DOC: update other MATLAB ode recommendation

eric-wieser committed:
BUG: Fix datetime->string conversion

size_t was being cast to char, causing bizarre errors like

    RuntimeError: The string provided for NumPy ISO datetime formatting was too short, with length 2000

Fixes #9712

eric-wieser committed:
TST: Use fresh state between tests

Old tests could be victim to stale data causing erroneous passes.

eric-wieser committed:
BUG: Allow subclasses of MaskedConstant to behave as unique singletons

Fixes astropy/astropy#6645

charris committed:
MAINT: Refactor rk_zipf to be more efficient.

Commit:
BUG: Added exception for casting numpy.ma.masked to long
TST: Added test_coercion_long and removed knownfailure from test_coercion_int

charris committed:
DOC: Redo some examples of np.arg(sort|max|min)

[ci skip]

ozabluda committed:
DOC: Improve documentation of axis parameter in numpy.unpackbits (#9828)

charris committed:
BUG: Check for NaN parameter in random.zipf.

Avoids infinite loop.

Commit:
BENCH: Added missing ufunc benchmarks
Added divmod, float_power, heaviside, is_nat and positive

Commit:
BUG: Raise TypeError instead of ValueError for ufunc isnat if operand is not of type datetime or timedelta
TST: Edited test_datetime to expect a TypeError instead of a ValueError for isnat if a wrong type is passed as an argument
DOC: Added change of ufunc isnat's exception handling to release notes

MSeifert04 committed:
DOC: ndarray.__copy__ takes no arguments [skip ci]

bashtage committed:
BUG: Prevent invalid array shapes in seed

Prevent empty arrays or arrays with more than 1 dimension from being
used to seed RandomState

closes #9832

eric-wieser committed:
BUG: Make bool(void_scalar) and void_scalar.astype(bool) consistent

Caused by void scalars decaying to 1d uint8 arrays before casting - `getitem` is dangerous for intermediate results

Works towards #9847

eric-wieser committed:
MAINT: Fix all special-casing of dtypes in `count_nonzero`

A quick profile reveals that `int.astype(bool)` is faster than `int == 0` anyway

xoviat committed:
BLD: distutils: auto-find vcpkg include and library directories

MSeifert04 committed:
DOC: Some minor fixes regarding import_array

[skip ci]

xoviat committed:
MAINT: make vcpkg include directories more clear

serhiy-storchaka committed:
BUG: Restore the environment variables when import multiarray fails.

The environment variables are modified before importing multiarray
and restored when the import is done.  But when the import fails
the environment varibles were left changed.

serhiy-storchaka committed:
ENH: Save to ZIP files without using temporary files.

Since Python 3.6 it is possible to write directly to a ZIP file,
without creating temporary files.

MSeifert04 committed:
DOC: Replace PyFITS reference with Astropy and PyTables with h5py

[skip ci]

eric-wieser committed:
BUG: Fix runtests --benchmark-compare in python 3

`check_output` returns bytes, but the rest of the command line is unicode.

Commit:
DOC: Update arraypad to use np.pad in examples

charris committed:
DOC: Make qr options render correctly as list.

[ci skip]

eric-wieser committed:
BUG: count_nonzero treats empty axis tuples strangely

Fixes #9728

This bug was introduced with the `axis` keyword in #7177, as a misguided optimization.

eric-wieser committed:
ENH: Implement ndarray.__format__ for 0d arrays

This fixes gh-7978

The behavior for other sized arrays is left unchanged, pending discussion in gh-5543

eric-wieser committed:
ENH: Improve unravel_index error message for out of bound values

eric-wieser committed:
BUG: Allow `unravel_index(0, ())` to return ()

Fixes gh-580

tovrstra committed:
DOC: Refine SVD documentation (#9845)

charris committed:
STY: Add blank line after variable declarations.

ahaldane committed:
BUG: add.reduce gives wrong results for arrays with funny strides

Fixes #9876

eric-wieser committed:
MAINT: Remove workarounds for gh-9527

The np.generic.astype method now accepts these keyword arguments

eric-wieser committed:
MAINT: Tidy np.histogram, and improve error messages

Split up the overloaded `bins` variable into separate names depending on its meaning

Helpful errors are now emitted for:
* non-integer bin counts (fixes gh-8072)
* non-1d bin edges

Removes another use of `np.isscalar`...

embray committed:
BUG: Allow __array_interface__['shape'] == () (#7994)

Previously this would fail when __array_interface__['data'] is a buffer object.

gh-2897 already removed this check for other types of data.

Commit:
ENH: Added compatability for the NAG Fortran compiler, nagfor

eric-wieser committed:
MAINT: rename mn and mx to first_edge and last_edge

andu97 committed:
TST: Improved tests for nag fortran compiler

adeak committed:
DOC: Unindent enumeration in savetxt docstring

The rendered markdown in the online documentation was broken due to
the one-character indentation added in the multiline enumerations of
the docstring of savetxt.

eric-wieser committed:
MAINT: Remove unused isscalar import

eric-wieser committed:
DOC: Fix incorrect comments that claim 0d arrays fail

charris committed:
MAINT: Punctuate `fromstring` docstring.

[ci skip]

charris committed:
STY: Fix C++ style comment.

[ci skip]

andu97 committed:
MAINT: made sure both compiler strings are included in test

charris commented:
Oops, you missed on the rebase, probably did a merge instead. 250+ commits seems a bit much ...

theodoregoetz commented:
Yes, yes I did! :) Fixing now.

I was trying to follow the rebasing on master section in docs as close
as possible here

https://docs.scipy.org/doc/numpy-1.13.0/dev/gitwash/development_workflo
w.html

but something went wrong it seems. The workflow docs are quite good,
this is my mistake for sure.
-- 
Goetz

On Mon, 2017-10-23 at 21:02 +0000, Charles Harris wrote:
> Oops, you missed on the rebase, probably did a merge instead. 250+
> commits seems a bit much ...
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

theodoregoetz commented:
double oops. Might go nuclear and refork - seems easiest path at this point.

theodoregoetz commented:
I don't have permission to reopen a pull request. I fixed things on my branch so just reopening should be enough to get this PR in order. Alternatively, I could just create a new PR...

eric-wieser commented:
PRs can't be reopened once their attached branch is deleted
